refactor(baselines) Upgrade FedBN Baseline to new flwr format

maddox-j opened the pull request:
## Issue

### Description

This PR upgrades the existing `FedBN` baseline from the previous `flwr` format to the new version.
Specifically:

1. Removed hydra library dependency
2. Adapted existing configuration files into pyproject.toml format
3. Use of Flower ClientApp and ServerApp
4. Kept manual handling of dataset (discussion points at the end of the PR).
5. Saving of results to disk by overloading the Server class definition.
6. Update README.md with steps to run baseline simulation through flwr run .
7. Result visualisation with Jupyter Notebook.
8. Ensured stateful clients through using the `ParametersRecord` in the client `Context`.

### Related issues/PRs

#4952 contains a related discussion about leaving the manual handling of the dataset.

## Proposal

### Explanation

This PR upgrades the existing `FedBN` baseline from the previous `flwr` format to the new version.

### Checklist

- [x] Implement proposed change
- [x] Write tests
- [x] Update [documentation](https://flower.ai/docs/writing-documentation.html)
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.ai/join-slack/) (channel `#contributions`)

### Any other comments?
N/A

jafermarq commented on baselines/fedbn/pyproject.toml:
```diff
@@ -141,3 +109,40 @@ exclude = [
+[tool.flwr.app.config]
+algorithm-name = "FedBN"
+num-server-rounds = 10
+num-clients = 5 # This is here so that we can access the number of clients during training. Does not change the actual amount of clients.
```
I believe this isn't needed since in the `Context` available in the `client_fn()` carries the information about the total number of nodes in the simulation. You can get that info like this:
```python
# num-partitions is a setting added at the start of the simulation
num_partitions = context.node_config["num-partitions"]
```

Worth trying if this works for your `get_data()` function?

maddox-j replied:
I think we still need to keep this because of the specification of the strategy in `server_app.py`

jafermarq reviewed (commented):
Hi @maddox-j, thank you for your patience for the first review. I left below some comments. I wasn't able to run the code because one `omegaconf` seemed not to be included as a dependency. If i install it, then some errors appear on the `ClientApp` side. : `KeyError: 'local_batch_norm'`.

I tested this with Flower 1.16.0 -- some small changes will be needed in those places where the ClientApp access the context if you want to update it to 1.17.0 since some classes and attributes were renamed. We can do that later.

maddox-j commented:
@jafermarq Thanks for the review. I addressed all the comments you had and made all the changes in one commit. This was all tested on `flwr=1.16.0`. Let me know what I should change to upgrade it to `1.17.0`.

jafermarq commented on baselines/fedbn/conf/fedavg.toml:
```diff
@@ -0,0 +1 @@
+algorithm_name = "FedAvg"
```
```suggestion
algorithm-name = "FedAvg"
```

jafermarq commented on baselines/fedbn/docs/multirun_plot.ipynb:
in `process_multirun_data()`, it should say `algorithm-name` instead of `algorithm_name`

jafermarq reviewed (commented):
Hey @maddox-j , it looks like the new `FedBN` is in a pretty good state! I was able to re-produce your results. Some suggestions are below.

maddox-j commented:
@jafermarq thanks! Suggestions incorporated 😎

jafermarq commented:
> @jafermarq thanks! Suggestions incorporated 😎

Great! I think the only remaining bits is to update it so it runs with `Flower 1.17.0` (the latest). From what i can see only the `client_app.py` needs some tweaks: (1) replacing `ParametersRecord` with `ArrayRecord`; and consequently (2) replace `client_state.parameters_records` with `client_state.array_records`.

maddox-j commented:
@jafermarq Nice! Done!

jafermarq commented on baselines/fedbn/pyproject.toml:
```diff
@@ -1,101 +1,68 @@
-    "Topic :: Software Development :: Libraries :: Python Modules",
-    "Typing :: Typed",
+dependencies = [
+    "flwr[simulation]>=1.16.0",
```
```suggestion
    "flwr[simulation]==1.17.0",
```
